prime-router/docs/proposals/0019-website-api-payload-structure.md

```diff
@@ -0,0 +1,332 @@
+    "reportItemCount" : 30,
+    "fileName" : "az-covid-19-e1676dcf-ddf7-4d22-aa11-13c8a81f0318-20230329195207.csv",
+    "fileType" : "CSV"
+}]
+```
+
+## Proposed changes
+
+### HTTP response codes
+
+The following is a non-exhaustive list of HTTP response codes that should commonly be used for the ReportStream website.  This should be a rough guideline of which HTTP response code to return from the API for all future endpoints.
+
+| Code | Name                  | Explanation                                                                                      | Example                                                            
+|------|-----------------------|--------------------------------------------------------------------------------------------------|--------------------------------------------------------------------| 
+| 200  | OK                    | Success with content                                                                             | Admin successfully requests settings for a particular Organization |
+| 201  | Created               | Success with new resource created                                                                | Admin successfully creates a new Sender/Receiver                   |
+| 400  | Bad Request           | Error; user submitted invalid/malformed data                                                     | Admin submits malformed Sender/Receiver settings data              
+| 401  | Unauthorized          | Error; user is not currently authorized to access the resource (may need to re-authenticate)     | Receiver requests deliveries with outdated Okta token              
+| 403  | Forbidden             | Error; user is never authorized to access the resource (re-authentication won't affect anything) | Receiver requests another Receiver's deliveries                    
+| 404  | Not Found             | Error; requested resource does not exist                                                         | Admin requests Receivers for a non-existent Organization           |
+| 500  | Internal Server Error | Unhandled exception                                                                              | Catch-all for all unhandled exceptions                             |
+
+NOTE: This table does not include common HTTP response codes commonly used as navigation signals, such as `301 Moved Permanently` or `308 Permanent Redirect`, which are not currently applicable to the website due to its single-page application (SPA) approach.  This may change in the future if we migrate to a more traditional website architecture.
+
+### HTTP methods
+
+There are no proposed changes to how we're using HTTP methods at the moment.  However, for all mutating methods (`POST`, `PUT`, `PATCH`, `DELETE`), we should return the entire modified record with all data through a `200 OK` rather than subsets of data or a `204 No Content`.  This is slightly atypical behavior, but it helps simplify request bookkeeping on the website.
```

Comment:
I think we should follow the spec wherever possible, and I don't think "request bookkeeping" is a strong enough reason to deviate from well-established practices? For example, the [latest spec](https://www.rfc-editor.org/rfc/rfc9110#name-post) I can find explains what a POST should return:


> If one or more resources has been created on the origin server as a result of successfully processing a POST request, the origin server SHOULD send a [201 (Created)](https://www.rfc-editor.org/rfc/rfc9110#status.201) response containing a [Location](https://www.rfc-editor.org/rfc/rfc9110#field.location) header field that provides an identifier for the primary resource created ([Section 10.2.2](https://www.rfc-editor.org/rfc/rfc9110#field.location)) and a representation that describes the status of the request while referring to the new resource(s).


Let me know if we should get together to discuss this further

Reply:
Ah, you're right; `201 Created` is correct for resource creation via `POST`, and I don't think we should deviate from this so I'll fix that.  My emphasis here was more about returning the entire record rather than subset of data specifically for `PUT`/`PATCH`/`DELETE`.  If it's a strong preference to not do this from the Platform team, that's fine -- it's only a small benefit to the WebX team and we can work around this.